### DISCUSSION TITLE
how to change complete provider to have below effect

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1524] [Q&A] [answered]
how to change complete provider to have below effect,now it seems that i can only write `v1.S` ,then it complete behind content `SET`,i want to achieve like v1. then behind content `SET` is remind complete
![image](https://github.com/eclipse-langium/langium/assets/26995040/10d6894b-8d7c-4cd9-86f2-bf29b0961f94)

**Comment by cdietrich** [comment]
am not sure if i understand your problem

- are you seeing no completions at `v1.|`.
- are you referring to Q1 and RESET1 should not be proposed and SET be auto applied?

**Reply by zrwsmd** [reply]
when i write v1. then all three of this element should be auto applied

**Comment by zrwsmd** [comment]
now i need to write v1.S,then v1.SET auto completed,
v1.Q ,then v1.Q1 auto completed,
v1.R,then v1.RESET1 auto completed,
for users,which seems not so intelligent,
that is ,as sonn as i write `.` then it should show `Q1`,`SET`,`RESET1` three elements

**Reply by msujew** [reply]
You need to set the conpletion options see:

https://github.com/eclipse-langium/langium/blob/6971c83afb924336d70b3e3092440478c73ac92e/packages/langium/src/lsp/completion/completion-provider.ts#L117

**Comment by zrwsmd** [comment]
just add this option?seems didn't work out
![image](https://github.com/eclipse-langium/langium/assets/26995040/6fbe63e4-cfe2-49d4-abc2-a6ee643e2882)

**Reply by msujew** [reply]
Why set the completions to undefined? This won't return any completions at all.

I think I misunderstood you initially. You want to automatically insert the recommended text, right? You need to asjust the `textsEdits` that are included in the `CompletionItem`.

**Reply by zrwsmd** [reply]
i mean for above `RS`,it actually a library function which user can call directly,it have three var decl,`Q1`,`SET`,`RESET1`,at moment,if i want to call it at st program,i decl a variable v1 at program var decl which is `RS` type,now is i write v1. ,it won't hint behind match var decl(`Q1`,`SET`,`RESET1`),i want to achieve like first screenshot,this effect is done by another software .now i can only achieve v1.Q | v1.S |v1.R then it hint behind content.for real user cases,user didn't know this library function what var elecment declar,so it need to write down v1.,then behind match var decl will show

**Reply by cdietrich** [reply]
I still don’t understand. Maybe you can provide a complete reproducer

**Reply by cdietrich** [reply]
Are you saying you don’t get any proposals at v1.

**Reply by msujew** [reply]
Ok, then using `triggerCharacters` should work as expected. Did you correctly bind it in your module? Also, don't use `undefined` for the `getCompletion` result as that will disable the completion.

**Reply by zrwsmd** [reply]
yeah,i bind it in my module correctly.so i need to override `getCompletion` method?i guess it  is a huge coding,
DefaultCompletionProvider&getCompletion can't be reuse?

**Reply by msujew** [reply, marked as answer]
No, don't override it at all. Just extend the `DefaultCompletionProvider` and set the `completionOptions`.

**Reply by zrwsmd** [reply]
it worked,thanks a lot

**Reply by zrwsmd** [reply]
`export class CacheCompletionProvider implements CompletionProvider {
    private defaultProvider: DefaultCompletionProvider;
    constructor(public services: LangiumServices) {
        this.defaultProvider = new DefaultCompletionProvider(services);
    }

    getCompletion(
        document: LangiumDocument<AstNode>,
        params: CompletionParams,
        cancelToken?: CancellationToken | undefined
    ): MaybePromise<CompletionList | undefined> {
        return this.defaultProvider.getCompletion(document, params);
    }

    readonly completionOptions: CompletionProviderOptions = {
        triggerCharacters: ['.']
        //allCommitCharacters: ['.']
    };
}`

**Reply by cdietrich** [reply]
why not simply extending DefaultCompletionProvider

**Reply by zrwsmd** [reply]
ok.i will change it
